File: FE 7-8 24-11 to/JS/24-01/abstraction.js
// Abstraction -> Hiding the method implementation details and it forces child classes to override parent class abstract methods
// abstract method --> method without implementation

class Payment {
    // abstract method
    pay(amount) {
        throw new Error("pay() method should be implemented in child class");
    }
}

class UPIPayment extends Payment {
    // method to pay amount
    pay(amount) {
        console.log(`Amount paid ${amount} using UPI...`);
    }
}

class CardPayment extends Payment {
    payy(amount) {
        console.log(`Amount paid ${amount} using Card...`);
    }
}

// function for processing the payment with given payment mode and given amount
function process_payment(payment_mode, amount) {
    // pay the amount using selected payment mode 
    payment_mode.pay(amount);
}


process_payment(new UPIPayment(), 1000); // paying amount using UPI
process_payment(new CardPayment(), 1200); // paying amount using Card

File: Python 6-7 10-12 to 29-01/24-01/abstraction.py
# Abstraction --> Hiding the method implementation details and provides standard structure to all child classes
# For abstraction need to inherit ABC class and use @abstractmethod for methods

from abc import ABC, abstractmethod

class Payment(ABC):
    # abstract method (method without any implementation)
    @abstractmethod
    def pay(self, amount):
        pass

class UPIPayment(Payment):
    # method overriding
    def pay(self, amount):
        print(f"Paid {amount} using UPI")

class CardPayment(Payment):
    # method overriding
    def pay(self, amount):
        print(f"Paid {amount} using credit card..")

# Adding new payment mode
class NetBankingPayment(Payment):
    def pay(self, amount):
        print(f"Amount paid {amount} using net banking...")


u1 = UPIPayment()
# u1.pay(500) # paying amount using UPI

c1 = CardPayment()
# c1.pay(500) # paying amount using Card

# function to process payemnt with selected payment mode
def process_payment(payment_mode, amount):
    # paying amount using selected payment mode
    payment_mode.pay(amount)


process_payment(UPIPayment(), 1000) # paying amount using UPI
process_payment(CardPayment(), 800) # paying amount using Card
process_payment(NetBankingPayment(), 1500) # paying amount using Card
